torch/_inductor/codegen/triton.py

```diff
@@ -727,6 +744,39 @@ def update_on_args(self, name, args, kwargs):
                 self.mask_vars.update({f"{arg.name[0]}mask"})
 
 
+def maybe_upcast_float32(convert_output: bool = True):
+    """
+    Codegen helper to upcast arguments to float32, depending on the config and dtype.
+    This decorates tl.math/libdevice codegen functions.
+    """
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        # Record that this function only supports float32 and float64.
+        OpDtypeSupport.register_upcast(func, convert_output)
+
+        def wrapped(*args, **kwargs) -> str:
+            # Optionally upcast args to float32.
+            dtype = args[0].dtype
```

Comment:
Some of these take in multiple inputs - can we iterate over args, kwargs ? in case args[1] is fp16, etc

Reply:
Good point. I originally wrote this only for unary ops, but now the PR is supposed to cover binary as well. I'll add a test case for a binary op where only one input needs to be upcast.

Reply:
I added the logic and a corresponding test in [77ee9a3](https://github.com/pytorch/pytorch/pull/140864/commits/77ee9a3c3c1c17e6e782e3e83c5d423092dbf2d5). I struggled to find a good test case, but the logic seems to make sense and doesn't break anything. At least for the cases I tried, it seems like inductor will insert casts at the graph level anyways so input tensors to tl.math/libdevice always end up having the same dtype by the time we get to codegen.